scripts/cnode-helper-scripts/cncli.sh

```diff
@@ -799,8 +804,271 @@ cncliPTsendslots() {
+  > "$tmpcsv" ; > "$csvfile" ; > "$onerow_csv"
+
+allEpochsInBlocklog() {
+  EPOCHS=$(sqlite3 "$BLOCKLOG_DB" "SELECT replace(group_concat(DISTINCT epoch ORDER BY epoch ASC), ',', ' ') AS EPOCHS FROM blocklog;")
+}
+
+getKoiosDataHist() {
+  [[ -z ${KOIOS_API} ]] && return 1
+
+  if ! pool_hist=$(curl -sSL -f "${KOIOS_API}/pool_history?_pool_bech32=${POOL_ID_BECH32}&_epoch_no=${EPOCH}" 2>&1); then
+    echo "ERROR: Koios pool_stake_snapshot history query failed."
+    return 1
+  fi
+
+  if ! epoch_hist=$(curl -sSL -f "${KOIOS_API}/epoch_info?_epoch_no=${EPOCH}" 2>&1); then
+    echo "ERROR: Koios epoch_stake_snapshot history query failed."
+    return 1
+  fi
+
+  pool_stake_hist=$(jq -r '.[].active_stake' <<< "${pool_hist}")
+  active_stake_hist=$(jq -r '.[].active_stake' <<< "${epoch_hist}")
+
+  return 0
+}
+
+checkEpochSlot() {
+  getNodeMetrics
+  local func=$1
+  local epochSlot=${slot_in_epoch}
+
+  if [[ $epochSlot -ge 0 && $epochSlot -le 21600 ]]; then
+    getCurrNextEpoch
+    echo "First six hours of new epoch "$curr_epoch" detected, too early to run this process."
+    echo "Current epochSlot is ${epochSlot}, please try againg after epochSlot 21600"
+    echo
+  else
+    $func
+  fi
+}
+
+getCurrNextEpoch() {
+   getNodeMetrics
+   curr_epoch=${epochnum}
+   next_epoch=$((curr_epoch + 1))
+}
+
+runCurrentEpoch() {
+  getKoiosData
+  echo "Processing current epoch: ${EPOCH}"
+  stake_param_curr="--active-stake ${active_stake_set} --pool-stake ${pool_stake_set}"
+
+  ${CNCLI} leaderlog ${cncliParams} --consensus "${consensus}" --epoch="${EPOCH}" --ledger-set current ${stake_param_curr} |
+      jq -r '[.epoch, .epochNonce, .poolId, .sigma, .d, .epochSlotsIdeal, .maxPerformance, .activeStake, .totalActiveStake] | @csv' |
+      sed 's/"//g' >> "$tmpcsv"
+}
+
+runNextEpoch() {
+  getKoiosData
+  echo "Processing next epoch: ${EPOCH}"
+  stake_param_next="--active-stake ${active_stake_mark} --pool-stake ${pool_stake_mark}"
+
+  ${CNCLI} leaderlog ${cncliParams} --consensus "${consensus}" --ledger-set next ${stake_param_next} |
+      jq -r '[.epoch, .epochNonce, .poolId, .sigma, .d, .epochSlotsIdeal, .maxPerformance, .activeStake, .totalActiveStake] | @csv' |
+      sed 's/"//g' >> "$tmpcsv"
+}
+
+runPreviousEpochs() {
+  getKoiosDataHist
+  echo "Processing previous epoch: ${EPOCH}"
+  stake_param_prev="--active-stake ${active_stake_hist} --pool-stake ${pool_stake_hist}"
+
+  ${CNCLI} leaderlog ${cncliParams} --consensus "${consensus}" --epoch="${EPOCH}" --ledger-set prev ${stake_param_prev} |
+      jq -r '[.epoch, .epochNonce, .poolId, .sigma, .d, .epochSlotsIdeal, .maxPerformance, .activeStake, .totalActiveStake] | @csv' |
+      sed 's/"//g' >> "$tmpcsv"
+}
+
+chooseConsensus() {
+  if (( EPOCH <= 363 )); then
+      consensus="tpraos"
+  elif (( EPOCH >= 364 )); then
+      consensus="praos"
+  else
+      consensus="cpraos"
+  fi
+}
```

Comment:
hardcoded for mainnet, fetch protocol version from Koios for epoch in question instead and use existing getConsensus function.
I propose the following changes:
- change cncli.sh getConsensus function and add a check if epoch is provided as argument to function.
- If so, call koios epoch_params function instead to get major and minor version for that epoch. Else, call normal getProtocolParams function.